### PR TITLE
feat: Improved focus management for toast stack

### PR DESCRIPTION
**Pull request by ling1726**
Adds some extra handling for multiple toasts:

* Arrow key navigation between toasts
* Focus trap within toaster
* Focus trap within toasts
* By default, focus will try to restore first to the most recent visible toast
* Escape will dismiss all toasts and restore focus
* Delete will dismiss the focused toast and restores focus
* Aria live announcements are disabled when the user is focused inside the toaster

Focus restoration after dismissing still works in the same way as before.

The focus management is a custom implementation since there is no way for tabster mover to work within a groupper microsoft/tabster#299

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/c6d2ace26c19b4c8102c77baf8f66ffeb36149f8'>c6d2ace26c19b4c8102c77baf8f66ffeb36149f8</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)

  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/FluentProviderWithTheme-mount_base.html">79</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/FluentProviderWithTheme-mount.html">79</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/FluentProviderWithTheme-mount.regression.txt">Possible regression</a></td>
           </tr></table><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/Avatar-mount_base.html">604</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/Avatar-mount.html">572</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/Button-mount_base.html">302</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/Button-mount.html">291</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/Field-mount_base.html">1028</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/Field-mount.html">1038</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/FluentProvider-mount_base.html">653</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/FluentProvider-mount.html">616</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/FluentProviderWithTheme-mount_base.html">79</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/FluentProviderWithTheme-mount.html">79</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/FluentProviderWithTheme-mount.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">72</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">69</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">74</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">72</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>InfoButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/InfoButton-mount_base.html">7</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/InfoButton-mount.html">10</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/MakeStyles-mount_base.html">850</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/MakeStyles-mount.html">857</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/Persona-mount_base.html">1634</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/Persona-mount.html">1562</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/SpinButton-mount_base.html">1320</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28660/perf-test-react-components/SpinButton-mount.html">1309</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-tabster/src/hooks/useFocusableGroup.ts [hunk 1]
```diff
@@ -7,6 +7,11 @@ export interface UseFocusableGroupOptions {
    * Behavior for the Tab key.
    */
   tabBehavior?: 'unlimited' | 'limited' | 'limited-trap-focus';
+
+  /**
+   * Tabster should ignore default handling of keydown events
```
Comment (layershifter) [inline packages/react-components/react-tabster/src/hooks/useFocusableGroup.ts:12]:
```suggestion
   * Tabster can ignore default handling of keydown events
```

?

File: packages/react-components/react-toast/src/components/Toast/Toast.cy.tsx [hunk 2]
```diff
@@ -565,21 +694,36 @@ describe('Toast', () => {
...
 
     mount(<Example />);
-
     cy.get('#make')
       .click()
-      .get('#dismiss')
-      .focus()
       .click()
-      .get(`.${toastClassNames.root}`)
+      .click()
```
Comment (layershifter) [inline packages/react-components/react-toast/src/components/Toast/Toast.cy.tsx:706]: Double click?

File: packages/react-components/react-toast/src/components/Toaster/useFocusManagement.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        acceptNode(node: Node) {
+          if (isHTMLElement(node) && node.classList.contains(toastContainerClassNames.root)) {
+            return NodeFilter.FILTER_ACCEPT;
+          }
+
+          return NodeFilter.FILTER_SKIP;
+        },
+      });
+
+      el.addEventListener('keydown', e => {
```
Comment (layershifter) [inline packages/react-components/react-toast/src/components/Toaster/useFocusManagement.ts:31]: I don't see a matching `.removeListener()` call. Can we clean up it properly?

File: packages/react-components/react-toast/src/components/Toaster/useFocusManagement.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * @internal
+ * Toasts  should be arrow navigable and focus should be trapped in a stack of tasts
+ * This is a temporary measure, Tabster does not have an API yet to enable mover arrow keys from within grouppers
+ * Once tabster fully supports this use case, remove this hook
+ */
+export function useFocusManagement_unstable() {
+  const { targetDocument } = useFluent();
+  return React.useCallback(
```
Comment (layershifter) [inline packages/react-components/react-toast/src/components/Toaster/useFocusManagement.ts:15]: Can we use `useEffect` there?

File: packages/react-components/react-toast/src/components/Toaster/useFocusManagement.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import { ArrowDown, ArrowUp } from '@fluentui/keyboard-keys';
+import { toastContainerClassNames } from '../ToastContainer';
+
+/**
+ * @internal
+ * Toasts  should be arrow navigable and focus should be trapped in a stack of tasts
+ * This is a temporary measure, Tabster does not have an API yet to enable mover arrow keys from within grouppers
+ * Once tabster fully supports this use case, remove this hook
+ */
+export function useFocusManagement_unstable() {
```
Comment (layershifter) [inline packages/react-components/react-toast/src/components/Toaster/useFocusManagement.ts:13]: Can you please prefix name with `Toaster` i.e. `useToaster`, so if someone will export this accidentally we will not have naming collisions.

File: packages/react-components/react-toast/src/components/Toaster/useFocusManagement.ts (hunk 3 above)
Comment (ling1726) [reply packages/react-components/react-toast/src/components/Toaster/useFocusManagement.ts]: Done in https://github.com/microsoft/fluentui/pull/28660/commits/4d12d8dbe0e2e7b617af8fe810f94df08c8a3896

**Review by layershifter — APPROVED** [review]

**Pull request merged**